crates/ruff_python_index/src/fstring_ranges.rs

```diff
@@ -0,0 +1,61 @@
+use std::collections::BTreeMap;
+
+use ruff_python_parser::Tok;
+use ruff_text_size::{TextRange, TextSize};
+
+/// Stores the ranges of all f-strings in a file sorted by [`TextRange::start`].
+/// There can be multiple overlapping ranges for nested f-strings.
+#[derive(Debug)]
+pub struct FStringRanges {
+    raw: BTreeMap<TextSize, TextRange>,
+}
+
+impl FStringRanges {
+    /// Return the [`TextRange`] of the innermost f-string at the given offset.
+    pub fn innermost(&self, offset: TextSize) -> Option<TextRange> {
+        self.raw
+            .range(..=offset)
+            .rev()
+            .find(|(_, range)| range.contains(offset))
+            .map(|(_, range)| *range)
```

Comment:
The worst-case complexity of this is that this function traverses all values in the map. Would it be sufficient to only look at the last element in the range?
```suggestion
				self.raw.range(..=offset).last().filter(|range| range.contains(offset)).copied()
```

It may be worth a comment why the filter is necessary

Reply:
> Would it be sufficient to only look at the last element in the range?

Unfortunately no. Let me explain using an example:

```python
f"first {f"second"} first"
```

Here, there are 2 f-strings with the following range:
1. `0..26` (outer)
2. `9..17` (inner)

Now, when we want to find the f-string for the offset 20 which is _after_ the inner f-string that's where the `find` comes in handy.

The `range` method will give both ranges because `range(..=20)` satisfies both. But, when iterating backwards, it isn't in (2) but in (1).

Reply:
Per your suggestion, we need to filter first and then get the last element:
```rust
        self.raw
            .range(..=offset)
            .filter(|(_, range)| range.contains(offset))
            .last()
            .map(|(_, range)| *range)
```

But, I think the original version is optimal as it'll short circuit on the first find while the filter would go through all the ranges and then get the last `Some` element.